### PR TITLE
Updated the computer player algorithm 

### PR DESCRIPTION
**Pull request by joshystuart**
## TLDR
I updated the computer player to use the minimax algorithm. It's impossible to beat it now:
X-tic-tac-toe-multiplayer-demo

## Background
After briefly looking over the code I noticed that it didn't have any tests that covered the react components. So initially I thought that it would be a good place to start ie. add a couple of very high level component tests so that if I added any new features, or refactored any code, I could be confident I didn't break anything. 

However, after spending some time trying to add react-testing-library (which required a jest upgrade, which in turn required babel updates and so on...), it quickly became obvious, that the amount of effort required would be more than just a few hours.

So with that in mind, I started a [TODO.md](https://github.com/joshystuart/X-ttt/blob/6e1b8dd06e130ccacbf65895fc3869763187a572/TODO.md) list to try to generate some ideas of what I would want to do if I had to look after this project going forward eg.

* uplift dependencies
* component unit tests
* functional tests
* linting/formatting/code conventions
* a simple CI pipeline
* some coding standards eg. separation of concerns, naming conventions, etc

Although these are NFRs, in my opinion they are essential for maintaining a large, complex project that is worked on by many people. And, despite it taking a little more time to implement these processes, it will certainly save a lot of effort in the long run by allowing us to deliver features much faster, with more confidence there won't be regression issues or bugs.

Since many of those ideas would result in massive code change sets (eg. linting/formatting...) -- which you suggested I avoid -- **I decided to try and enhance the way the computer chooses it's next move**.

## In this PR
Below are some of the high level tasks I did as part of this PR
* Updated the algorithm used for the computer player to choose its next move from a random one to a [minimax](https://en.wikipedia.org/wiki/Minimax) one.
* Added some test coverage for the new code.
* Added a `TODO` list for things I'd consider doing to improve the project.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 .svn/
+.idea
```
Comment (joshystuart) [inline .gitignore:2]: ignore the webstorm metadata

File: TODO.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# TODO
```
Comment (joshystuart) [inline TODO.md:1]: After failing to add component unit tests, I decided to make a list of things that would help improve this project, then I could select 1 or 2 for this PR. Unfortunately, because I spent so much time in dependency hell trying to get component unit tests working, I could only check off 1 of these.

File: react_ws_src/package-lock.json
Comment (joshystuart) [inline react_ws_src/package-lock.json:1]: Ideally there should be an [.nvm config](https://github.com/nvm-sh/nvm) file or [engines](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#engines) specified in `package.json` so the same node version is used to avoid regenerating the lock.

Also, there's a bug where if you run `npm install` multiple times, it will fail with a `could not resolve`:
image

Again, I tried updating dependencies, but because they are >7 years old, it was a game of "whack-a-mole", and I gave up.

You can work around the error using: `npm install --legacy-peer-deps`, but fixing the old dependencies would be much better.

File: react_ws_src/src/helpers/__tests__/computer_moves-test.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+jest.unmock('../computer_moves');
```
Comment (joshystuart) [inline react_ws_src/src/helpers/__tests__/computer_moves-test.js:1]: Im not sure why I had to unmock the file? And I couldn't find any documentation on `jest v13`. It seemed to work, so I left it.

File: react_ws_src/src/helpers/__tests__/computer_moves-test.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import {
+  convert_cell_to_index,
+  convert_index_to_cell,
+  convert_move_coordinates_to_move_position_array,
+  get_available_moves,
+  get_board, get_computer_move,
+  get_minimax,
+  is_winner,
+  normalize_move_coordinates,
+} from '../computer_moves';
```
Comment (joshystuart) [inline react_ws_src/src/helpers/__tests__/computer_moves-test.js:11]: Ideally I wouldn't have exported all the internal functions (in OO, they'd be considered `private`), but I think it was better to break things down and test them individually rather than trying to handle all the permutations via the main `get_computer_move` function.

File: react_ws_src/src/helpers/computer_moves.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
+const BOARD_SIZE = 9;
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:1]: Although I haven't changed the board size, I remember you suggesting that could be one of the tasks I do. So although I'm not passing it into the `get_computer_move` function, or getting it from `config`, I have _parametized_ it so that it would be easier to do later.

File: react_ws_src/src/helpers/computer_moves.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+const CELL_PREFIX = 'c';
+const PLAYER = 'x';
+const COMPUTER = 'o';
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:4]: These would be better off being driven from some kind of `globals` or `config` so that they are consistent throughout the app, but I didn't want to start making the changeset even bigger. It could be done easily in another PR.

File: react_ws_src/src/helpers/computer_moves.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+/**
+ * Finds the best move for the computer using the minimax algorithm.
+ * It converts the move index into the cell id used within the UI.
+ *
+ * NOTE: ideally we wouldn't have to convert between the cell id and the index since that is a UI concern.
+ * Nevertheless, I felt it was better to encapsulate everything into 1 function. If I had more time I'd refactor
+ * the component a bit to create clearer separation of concerns between the move logic and the presentation.
+ */
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:13]: This is the main entry point. I prefer named exports over `defaults` because it means you retain the same function name everywhere.

As the `note` above suggests, ideally we would not be passing in data that is coupled to the UI. However, it was cleaner to handle the conversion from `cell` to `index` and back here, rather than in the component, at least for now. It allowed me to test it more thoroughly and be more confident it would work.

File: react_ws_src/src/helpers/computer_moves.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+export function get_board(size) {
+  return Array(size).fill().map((x, i) => i);
+}
+
+/**
+ * Given the current moves by each player applies them to the board array
+ */
+export function convert_move_coordinates_to_move_position_array(current_moves) {
+  const board = get_board(BOARD_SIZE);
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:59]: Even though I parametized the board size, I didn't put any other safe guards around it. ie. we should only allow odd numbers with perfect squares, otherwise the game becomes unplayable. I figured since the board size is hardcoded for now, it's not a big deal.

File: react_ws_src/src/helpers/computer_moves.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+ * Returns only the positions on the board that can be played
+ */
+export function get_available_moves(board) {
+  return board.filter((position) => [PLAYER, COMPUTER].indexOf(position) === -1);
+}
+
+/**
+ * Checks if the current ttt board (of any size) has a winner.
+ */
+export function is_winner(board, player) {
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:78]: This function could have been much simpler if I didn't account for the variable board size.

File: react_ws_src/src/helpers/computer_moves.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  for (let i = 0; i < size; i++) {
+    // check the diagonals
+    diag1 = diag1 && (board[i * size + i] === player);
+    diag2 = diag2 && (board[i * size + (size - i - 1)] === player);
+  }
+
+  return diag1 || diag2;
+}
+
+export function get_minimax(board, player) {
```
Comment (joshystuart) [inline react_ws_src/src/helpers/computer_moves.js:107]: This is the magic that calculates where to move based on: https://en.wikipedia.org/wiki/Minimax

File: react_ws_src/src/views/ttt/GameMain.js [hunk 11]
```diff
@@ -1,11 +1,10 @@
-import React, {Component} from 'react'
```
Comment (joshystuart) [inline react_ws_src/src/views/ttt/GameMain.js:1]: Sorry about the re-formatting: when I added the new `get_computer_move` function into this file, my IDE automatically cleaned up the imports.

File: react_ws_src/src/views/ttt/GameMain.js [hunk 12]
```diff
@@ -174,7 +173,7 @@ export default class SetName extends Component {
 
 		cell_vals[cell_id] = 'x'
 
-		TweenMax.from(this.refs[cell_id], 0.7, {opacity: 0, scaleX:0, scaleY:0, ease: Power4.easeOut})
+		this.animate_cell(cell_id);
```
Comment (joshystuart) [inline react_ws_src/src/views/ttt/GameMain.js:176]: I noticed `TweenMax.from(this.refs[cell_id], 0.7, {opacity: 0, scaleX:0, scaleY:0, ease: Power4.easeOut})` was copy/pasted a bunch here, so I thought to be a bit more DRY, I'd move it to a function. It's probably something that could be abstracted even more into some kind of `animation_helper.js`

File: react_ws_src/src/views/ttt/GameMain.js [hunk 13]
```diff
@@ -192,29 +191,15 @@ export default class SetName extends Component {
 //	------------------------	------------------------	------------------------
 
 	turn_comp () {
```
Comment (joshystuart) [inline react_ws_src/src/views/ttt/GameMain.js:193]: I wanted to make minimal changes to this file, which is why I've encapsulated all of the logic into `const cell_id = get_computer_move(cell_vals);`

Everything else, except the animation and state mutation, could be removed in this function.

File: react_ws_src/src/views/ttt/GameMain.js [hunk 14]
```diff
@@ -339,5 +324,9 @@ export default class SetName extends Component {
 	}
 
 
-
+	animate_cell(cell_id) {
```
Comment (joshystuart) [inline react_ws_src/src/views/ttt/GameMain.js:327]: This could use some formatting, but I didn't want to reformat the whole file. IMO the project should have `prettier` that runs on commit and in CI as a check. It would help clean up the formatting changesets and keep everything consistent.
